pkg/pattern/drain/line_tokenizer.go

```diff
@@ -1,25 +1,96 @@
 package drain
 
-import "strings"
+import (
+	"strings"
+	"unicode"
+)
 
 type LineTokenizer interface {
-	Tokenize(line string) []string
-	Join(tokens []string) string
+	Tokenize(line string) ([]string, interface{})
+	Join(tokens []string, state interface{}) string
 }
 
 type spacesTokenizer struct{}
 
-func (spacesTokenizer) Tokenize(line string) []string {
-	return strings.Split(line, " ")
+func (spacesTokenizer) Tokenize(line string) ([]string, interface{}) {
+	return strings.Split(line, " "), nil
 }
 
-func (spacesTokenizer) Join(tokens []string) string {
+func (spacesTokenizer) Join(tokens []string, _ interface{}) string {
 	return strings.Join(tokens, " ")
 }
 
+type punctuationTokenizer struct {
+	includeDelimiters [128]rune
+	excludeDelimiters [128]rune
+}
+
+func newPunctuationTokenizer() *punctuationTokenizer {
+	var included [128]rune
+	var excluded [128]rune
+	included['='] = 1
+	excluded['_'] = 1
+	excluded['-'] = 1
+	return &punctuationTokenizer{
+		includeDelimiters: included,
+		excludeDelimiters: excluded,
+	}
+}
+
+func (p *punctuationTokenizer) Tokenize(line string) ([]string, interface{}) {
+	tokens := make([]string, len(line))                  // Maximum size is every character is punctuation
+	spacesAfter := make([]int, strings.Count(line, " ")) // Could be a bitmap, but it's not worth it for a few bytes.
```

Comment:
You might want to use a pool for this one. Prometheus has a good sync.Pool that works in buckets.